README.md
# Scala Exercises - Functional Programming Principles library
------------------------

This repository hosts a library for the first course of the [Scala MOOC](https://www.coursera.org/specializations/scala) ("Functional Programming Principles in Scala").

## Run Locally

- Clone this repository, compile and publish the project:

~~~ sh
git clone git@github.com:scala-exercises/exercises-scalatutorial.git
cd exercises-scalatutorial/
sbt compile publishLocal # it is important to first run the `compile` command alone
~~~

- Clone the `evaluator` and run it:

~~~ sh
git clone git@github.com:scala-exercises/evaluator.git
cd evaluator/
sbt "project evaluator-server" run
~~~

- Clone the `scala-tutorial` branch of our `scala-exercises` fork:

~~~ sh
git clone -b scala-tutorial git@github.com:scalacenter/scala-exercises.git
~~~

- Follow the database setup instructions given
  [here](https://github.com/scala-exercises/scala-exercises#configure-the-database)

- Add the following line to `server/conf/application.dev.conf`:

~~~
evaluator.secretKey="secretKey"
~~~

- Run the server:

~~~ sh
sbt -mem 1500 run
~~~

## About Scala Exercises

"Scala Exercises" brings exercises for the Stdlib, Cats, Shapeless, and many other great libraries for Scala to your browser. It offers hundreds of solvable exercises organized into several categories covering the basics of the Scala language and its most important libraries.

Scala Exercises is available at [scala-exercises.org](https://scala-exercises.org).

## Contributing

Contributions welcome! Please join our [Gitter channel](https://gitter.im/scala-exercises/scala-exercises)
to get involved, or visit our [GitHub site](https://github.com/scala-exercises).

## License

Copyright (C) 2015-2019 47 Degrees, LLC.
Reactive, scalable software solutions.
http://47deg.com
hello@47deg.com

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# Changelog

## [v0.6.6](https://github.com/scala-exercises/exercises-scalatutorial/tree/v0.6.6) (2020-08-10)

[Full Changelog](https://github.com/scala-exercises/exercises-scalatutorial/compare/v0.6.4...v0.6.6)

📈 **Dependency updates**

- Update definitions, exercise-compiler, ... to 0.6.4 [\#154](https://github.com/scala-exercises/exercises-scalatutorial/pull/154) ([47erbot](https://github.com/47erbot))
- Update sbt-mdoc to 2.2.4 [\#153](https://github.com/scala-exercises/exercises-scalatutorial/pull/153) ([47erbot](https://github.com/47erbot))
- Update scalacheck-1-14 to 3.2.1.0 [\#152](https://github.com/scala-exercises/exercises-scalatutorial/pull/152) ([47erbot](https://github.com/47erbot))
- Update scalatest to 3.2.1 [\#151](https://github.com/scala-exercises/exercises-scalatutorial/pull/151) ([47erbot](https://github.com/47erbot))
- Update sbt-scalafmt to 2.4.2 [\#148](https://github.com/scala-exercises/exercises-scalatutorial/pull/148) ([47erbot](https://github.com/47erbot))
- Update scalafmt-core to 2.6.4 [\#147](https://github.com/scala-exercises/exercises-scalatutorial/pull/147) ([47erbot](https://github.com/47erbot))
- Update scalafmt-core to 2.6.3 [\#146](https://github.com/scala-exercises/exercises-scalatutorial/pull/146) ([scala-steward](https://github.com/scala-steward))
- Update scalafmt-core to 2.6.2 [\#144](https://github.com/scala-exercises/exercises-scalatutorial/pull/144) ([scala-steward](https://github.com/scala-steward))

**Merged pull requests:**

- Remove double parenthesis in tuple exercise [\#149](https://github.com/scala-exercises/exercises-scalatutorial/pull/149) ([Gerko32](https://github.com/Gerko32))

## [v0.6.4](https://github.com/scala-exercises/exercises-scalatutorial/tree/v0.6.4) (2020-06-30)

[Full Changelog](https://github.com/scala-exercises/exercises-scalatutorial/compare/v0.6.3...v0.6.4)

## [v0.6.3](https://github.com/scala-exercises/exercises-scalatutorial/tree/v0.6.3) (2020-06-29)

[Full Changelog](https://github.com/scala-exercises/exercises-scalatutorial/compare/v0.6.2...v0.6.3)

📈 **Dependency updates**

- Update sbt to 1.3.13 [\#139](https://github.com/scala-exercises/exercises-scalatutorial/pull/139) ([scala-steward](https://github.com/scala-steward))
- Update sbt-mdoc to 2.2.3 [\#137](https://github.com/scala-exercises/exercises-scalatutorial/pull/137) ([scala-steward](https://github.com/scala-steward))
- Update scalacheck-1-14 to 3.2.0.0 [\#134](https://github.com/scala-exercises/exercises-scalatutorial/pull/134) ([scala-steward](https://github.com/scala-steward))
- Update scalatest to 3.2.0 [\#133](https://github.com/scala-exercises/exercises-scalatutorial/pull/133) ([scala-steward](https://github.com/scala-steward))
- Update sbt-mdoc to 2.2.2 [\#131](https://github.com/scala-exercises/exercises-scalatutorial/pull/131) ([scala-steward](https://github.com/scala-steward))
- Update sbt to 1.3.11 [\#127](https://github.com/scala-exercises/exercises-scalatutorial/pull/127) ([scala-steward](https://github.com/scala-steward))
- Update scalatest to 3.1.2 [\#126](https://github.com/scala-exercises/exercises-scalatutorial/pull/126) ([scala-steward](https://github.com/scala-steward))

**Merged pull requests:**

- Updates build [\#140](https://github.com/scala-exercises/exercises-scalatutorial/pull/140) ([juanpedromoreno](https://github.com/juanpedromoreno))
- Update scalafmt-core to 2.6.1 [\#135](https://github.com/scala-exercises/exercises-scalatutorial/pull/135) ([BenFradet](https://github.com/BenFradet))
- Updated SyntacticConveniences.scala [\#129](https://github.com/scala-exercises/exercises-scalatutorial/pull/129) ([aamirfayaz](https://github.com/aamirfayaz))
- Prepare repository for next  release and SBT build improvements [\#128](https://github.com/scala-exercises/exercises-scalatutorial/pull/128) ([juanpedromoreno](https://github.com/juanpedromoreno))

## [v0.6.2](https://github.com/scala-exercises/exercises-scalatutorial/tree/v0.6.2) (2020-04-27)

[Full Changelog](https://github.com/scala-exercises/exercises-scalatutorial/compare/v0.6.1...v0.6.2)

## [v0.6.1](https://github.com/scala-exercises/exercises-scalatutorial/tree/v0.6.1) (2020-04-27)

[Full Changelog](https://github.com/scala-exercises/exercises-scalatutorial/compare/v0.6.0...v0.6.1)

**Merged pull requests:**

- Compile when Publish [\#121](https://github.com/scala-exercises/exercises-scalatutorial/pull/121) ([juanpedromoreno](https://github.com/juanpedromoreno))

## [v0.6.0](https://github.com/scala-exercises/exercises-scalatutorial/tree/v0.6.0) (2020-04-24)

[Full Changelog](https://github.com/scala-exercises/exercises-scalatutorial/compare/v0.4.0...v0.6.0)

**Closed issues:**

- Scala Tutorial: the last exercise of Lazy Evaluation, builder.result\(\) shouldBe "xyzz"  [\#38](https://github.com/scala-exercises/exercises-scalatutorial/issues/38)
- Rename `aux` folder in `exercises-scalatutorial` content library, as its incompatible in Windows machines [\#28](https://github.com/scala-exercises/exercises-scalatutorial/issues/28)
- Syntax isn't introduced [\#23](https://github.com/scala-exercises/exercises-scalatutorial/issues/23)
- I can't press run button [\#15](https://github.com/scala-exercises/exercises-scalatutorial/issues/15)

**Merged pull requests:**

- Updates Project [\#116](https://github.com/scala-exercises/exercises-scalatutorial/pull/116) ([juanpedromoreno](https://github.com/juanpedromoreno))
- "arbitrary" -\> "arbitrarily" [\#114](https://github.com/scala-exercises/exercises-scalatutorial/pull/114) ([alexmdac](https://github.com/alexmdac))
- Update sbt to 1.3.10 [\#113](https://github.com/scala-exercises/exercises-scalatutorial/pull/113) ([scala-steward](https://github.com/scala-steward))
- Define nthPrime\(\) before called by secondPrime\(\) [\#110](https://github.com/scala-exercises/exercises-scalatutorial/pull/110) ([xiaowei-zhang](https://github.com/xiaowei-zhang))

project/ProjectPlugin.scala
import sbt.Keys._
import sbt._
import com.alejandrohdezma.sbt.github.SbtGithubPlugin

object ProjectPlugin extends AutoPlugin {

  override def trigger: PluginTrigger = allRequirements

  override def requires: Plugins = plugins.JvmPlugin && SbtGithubPlugin

  object autoImport {

    lazy val V = new {
      val shapeless: String           = "2.3.3"
      val scala: String               = "2.13.2"
      val scalaExercises: String      = "0.6.0"
      val scalacheck: String          = "1.14.3"
      val scalacheckShapeless: String = "1.2.5"
      val scalatest: String           = "3.1.2"
      val scalatestplusScheck: String = "3.1.1.1"
    }

    def dep(artifactId: String) = "org.scala-exercises" %% artifactId % V.scalaExercises

    lazy val exercisesSettings = Seq(
      libraryDependencies ++= Seq(
        dep("exercise-compiler"),
        dep("definitions"),
        "com.chuusai"                %% "shapeless"                 % V.shapeless,
        "org.scalatest"              %% "scalatest"                 % V.scalatest,
        "org.scalacheck"             %% "scalacheck"                % V.scalacheck,
        "org.scalatestplus"          %% "scalacheck-1-14"           % V.scalatestplusScheck,
        "com.github.alexarchambault" %% "scalacheck-shapeless_1.14" % V.scalacheckShapeless
      )
    )
  }

  override def projectSettings: Seq[Def.Setting[_]] =
    Seq(
      organization := "org.scala-exercises",
      organizationName := "47 Degrees",
      organizationHomepage := Some(url("https://47deg.com")),
      scalaVersion := autoImport.V.scala,
      resolvers ++= Seq(
        Resolver.sonatypeRepo("snapshots"),
        Resolver.sonatypeRepo("releases")
      )
    )
}
